coordinator/src/notifications.rs

```diff
@@ -0,0 +1,356 @@
+use crate::db::positions;
+use crate::db::user;
+use crate::position::models::Position;
+use anyhow::Context;
+use anyhow::Result;
+use diesel::PgConnection;
+use std::fmt::Display;
+use time::OffsetDateTime;
+use tokio::sync::mpsc;
+
+/// Consider a position expiring soon if it expires in less than this time
+const START_OF_EXPIRING_POSITION: time::Duration = time::Duration::hours(1);
+
+/// Consider a position expired if it expired more than this time ago
+const END_OF_EXPIRED_POSITION: time::Duration = time::Duration::hours(1);
+
+/// Types of notification that can be sent to 10101 app users
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum NotificationKind {
+    /// Coordinator would like to settle the channel
+    ChannelClose,
+    PositionSoonToExpire,
+    PositionExpired,
+}
+
+impl Display for NotificationKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NotificationKind::ChannelClose => write!(f, "ChannelClose"),
+            NotificationKind::PositionSoonToExpire => write!(f, "PositionSoonToExpire"),
+            NotificationKind::PositionExpired => write!(f, "PositionExpired"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Notification {
+    pub user_fcm_token: FcmToken,
+    pub notification_kind: NotificationKind,
+}
+
+impl Notification {
+    pub fn new(user_fcm_token: FcmToken, notification_kind: NotificationKind) -> Self {
+        Self {
+            notification_kind,
+            user_fcm_token,
+        }
+    }
+}
+
+/// Actor managing the notifications
+pub struct NotificationService {
+    notification_sender: mpsc::Sender<Notification>,
+}
+
+impl NotificationService {
+    /// Start the notification service
+    ///
+    /// If an empty string is passed in the constructor, the service will not send any notification.
+    /// It will only log the notification that it would have sent.
+    pub fn new(fcm_api_key: String) -> Self {
```

Comment:
I know it wasn't introduced here, but wouldn't it be nicer to use an option? 
The advantage of using an option is that we move the check into the type system that the value is set while if it is `empty` you as the dev have to check it.

Reply:
this one mentioned here  is only the API key (the one that it is likely an option, e.g. in local testing). 

In this PR I've introduced `FcmToken` type that checks during initialisation if it's not an empty string, so that we can take advantage of the type system (these ones should be populated, unless a person has disabled push notifications - haven't tested that theory, but I assume you would not get a firebase token if you said 'no' to notification permission).